File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.py
import sys

n = int(input())

g = [list(map(int, input().split())) for _ in range(n)]
visited = [0] * n # 방문 체크 용

res = int(1e9) # 최소 비용
def dfs(start, v, sum_data): # start는 시작 노드 기억을 위해. v는 현재 노드, sum은 지금까지의 합 !
    global res
    if sum_data > res:
        return # 가지치기
    if sum(visited) == n: # 모두 방문했어.
        if g[v][start] != 0: # 다시 시작노드로 돌아가야한다 -> 이런 조건은 기록해놔야해.
            sum_data += g[v][start] # 시작노드로 가는 비용도 더했어야 !!! 
            if sum_data < res:
                res = sum_data
        return
    else: # 더 가야해
        for i in range(n):
            if visited[i] == 0 and g[v][i] > 0: # 방문하지 않았으면서 갈 수 있어야해
                visited[i] = 1
                dfs(start, i, sum_data + g[v][i])
                visited[i] = 0 # 백트랙킹 시 원상 복구

for i in range(n):
    visited = [0] * n # 계속 만들어줘야지
    visited[i] = 1 # 현재 노드가 시작점이니 방문처리 후에
    dfs(i,i,0)
print(res)

File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static int[][] g;
    static int minDis = Integer.MAX_VALUE;
    static boolean[] visited;

    public static void DFS(int L, int sum, int v, int start) {
        if (L == n) { // ! n개의 도시를 모두 확인해서 돌아왔을 때
            int cnt = 0;
            for (int i = 1; i <= n; i++) {
                if (visited[i]) {
                    cnt += 1;
                }
            }
            if (cnt == n) { // 모두 방문한 경우
                if (g[v][start] != 0) {  // 시작점으로 다시 돌아올 수 있는 경우
                    minDis = Math.min(minDis, sum + g[v][start]);
                }
            }
        } else {
            for (int i = 1; i <= n; i++) {
                if (g[v][i] != 0 && !visited[i]) {  // 두 정점 연결되어 있는 경우
                    visited[i] = true;
//                    System.out.println("현재 노드 : " + v + " 현재 방문할 노드 : " + i);
                    DFS(L + 1, sum + g[v][i], i, start);
                    visited[i] = false;
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;


        n = Integer.parseInt(br.readLine());
        g = new int[n + 1][n + 1];
        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= n; j++) {
                g[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 1; i <= n; i++) {
            visited = new boolean[n + 1];
            visited[i] = true; // 시작지점 방문처리 후 출발
            DFS(1, 0, i, i);
        }

        System.out.println(minDis);

    }
}